llamabot/bot/pydanticbot.py

```diff
@@ -0,0 +1,50 @@
+import contextvars
+
+from llamabot import ChatBot
+from llamabot.components.messages import (
+    AIMessage,
+)
+from pydantic import BaseModel, ValidationError
+
+
+class PydanticBot(ChatBot):
+    """Pydantic Bot is given a Pydantic Model and expects the LLM to return a JSON structure that conforms to the model schema
+    It will validate the model, repeating the prompt if it does not validate, and then returns an instance of that model.
+    """
```

Comment:
I added a new docstring incorporating some of these changes.
And made more explicit why this is different to JSON mode. But essentially, my experience with JSON mode was that it only guarantees JSON parsable output, but it does not guarantee the schema we prompted for. This was especially true when prompting with a large context length, it kinda forgot what schema I asked it to use.
Then, ontop of ensuring the right schema, we can customize the validation rules. For example, if your extracting a name of a speaker from a large prompt, the validation can check that the name actually exist within the initial prompt, to avoid hallucinations.
This is could also be used for summarizing with citations to the original source. You could imagine a schema like this:

```
class SummarySentence(BaseModel):
    """ a single sentence as part of the summary of the given text """
    sentence: str = Field(description="A sentence summary of the given text")
    quote: str = Field(
        description="directly quoting the given text, where this summary was generated from")

   @field_validator('quote')
    def validate_name(cls, quote):
        """ check that actually exists in the given text, if it doesn't this could be a hallucination, if it does we can use the index of this quote as a citation to jump straight to the source. """
        pass

class Summary(BaseModel):
    """ a paragraph summarizing the given text """
    summary: List(SummarySentence)
```



I've also moved away from inheriting that behaviour from ChatBot, as I think inheriting from SimpleBot makes more sense.
Initially, like you said, I wanted to be able to submit the history for negative examples.
However, implementing that by overriding the call method and storing an array of attempts is simpler, and cleaner for the end user.

Reply:
Wow, @ElliotSalisbury that SummarySentence model is incredibly creative! I like the "quote" idea as it effectively enables citation from the text. I'm learning lots observing how you design the pydantic models!